sklearn/linear_model/sag_fast.pyx

```diff
@@ -32,49 +27,219 @@ cdef inline double fmax(double x, double y) nogil:
     return y
 
 
+cdef double _logsumexp(double* arr, int n_classes) nogil:
+    """Computes the sum of arr assuming arr is in the log domain.
+
+    Returns log(sum(exp(arr))) while minimizing the possibility of
+    over/underflow.
+    """
+    # Use the max to normalize, as with the log this is what accumulates
+    # the less errors
+    cdef double vmax = arr[0]
+    cdef double out = 0.0
+    cdef int i
+
+    for i in range(1, n_classes):
+        if vmax < arr[i]:
+            vmax = arr[i]
+
+    for i in range(n_classes):
+        out += exp(arr[i] - vmax)
+
+    return log(out) + vmax
+
+
+cdef class MultinomialLogLoss:
```

Comment:
```python
"""
The multinomial logistic loss is equal to:
loss = - w \sum_c \delta_{y,c} (prediction[c] - logsumexp(prediction))
     = w (logsumexp(prediction) - prediction[y])

where:
    prediction = dot(x, w) + intercept
    \delta_{y,c} = 1 if (y == c) else 0
"""
```
and 
```python
"""
The gradient of the multinomial logistic loss is equal to:
grad_c = - w * (p[c] - \delta_{y,c})

where:
    p[c] = exp(logsumexp(prediction) - prediction[c])
    prediction = dot(x, w) + intercept
    \delta_{y,c} = 1 if (y == c) else 0

Note that to obtain the true gradient, this value has to be multiplied
by the sample vector x.
"""
```
Feel free to suggest a clearer formulation, I don't like this one so much